### PR TITLE
Ingesters must flush chunks then remove themselves from consul!

### PR DESCRIPTION
**Pull request by tomwilkie**
Part of #427

### CODE REVIEW

File: pkg/ingester/ingester_lifecycle_test.go [hunk 1]
```diff
@@ -268,3 +270,68 @@ func (i ingesterClientAdapater) TransferChunks(ctx context.Context, _ ...grpc.Ca
...
+	_, err = ing.Push(ctx, util.ToWriteRequest([]model.Sample{
+		{
+			Metric:    m,
+			Timestamp: ts,
+			Value:     val,
+		},
+	}))
+	require.NoError(t, err)
+
+	// We add a 100ms sleep into the flush loop, such that we can reliably detect
```
Comment (jml) [inline pkg/ingester/ingester_lifecycle_test.go]: What is it about 100ms that makes it reliable? This kind of test is notoriously unreliable.

I think it's worth exploring a non-polling solution to this.

File: pkg/ingester/ingester_lifecycle_test.go [hunk 2]
```diff
@@ -268,3 +270,68 @@ func (i ingesterClientAdapater) TransferChunks(ctx context.Context, _ ...grpc.Ca
...
+	ing.flushUserSeries = func(userID string, fp model.Fingerprint, immediate bool) error {
+		time.Sleep(100 * time.Millisecond)
+		return ing.flushUserSeriesImpl(userID, fp, immediate)
+	}
+
+	// Now stop the ingester.  Don't call shutdown, as it waits for all goroutines
+	// to exit.  We just want to check that by the time the token is removed from
+	// the ring, the data is in the chunk store.
+	close(ing.quit)
+	poll(t, 200*time.Millisecond, 0, func() interface{} {
```
Comment (jml) [inline pkg/ingester/ingester_lifecycle_test.go]: Is this 200ms significant?

File: pkg/ingester/ingester_lifecycle_test.go [hunk 3]
```diff
@@ -268,3 +270,68 @@ func (i ingesterClientAdapater) TransferChunks(ctx context.Context, _ ...grpc.Ca
...
+		if err != nil {
+			return -1
+		}
+		return len(r.(*ring.Desc).Ingesters)
+	})
+
+	// And check the store has the chunk
+	res, err := chunk.ChunksToMatrix(store.chunks[userID])
+	require.NoError(t, err)
+	sort.Sort(res)
```
Comment (jml) [inline pkg/ingester/ingester_lifecycle_test.go]: Is this sort necessary? The `Matrix` has one element, the `SampleStream` is a struct, and the `Values` slice has only one `SamplePair`. What is being sorted?

File: pkg/ingester/ingester.go [hunk 4]
```diff
@@ -150,7 +150,11 @@ type Ingester struct {
 
 	// One queue per flush thread.  Fingerprint is used to
 	// pick a queue.
-	flushQueues []*util.PriorityQueue
+	flushQueues     []*util.PriorityQueue
+	flushQueuesDone sync.WaitGroup
+
+	// Virtual lookup for this function to allow the tests to get in there.
+	flushUserSeries func(userID string, fp model.Fingerprint, immediate bool) error
```
Comment (jml) [inline pkg/ingester/ingester.go]: I'd rather do this by leaving `flushUserSeries` as it was and instead having a `preFlushUserSeries` hook that is called by `flushUsersSeries`. It costs an extra function call each time, but it makes the structure of this object a bit more obvious.

File: pkg/ingester/ingester_lifecycle_test.go (hunk 3 above)
Comment (tomwilkie) [inline pkg/ingester/ingester_lifecycle_test.go]: Just copy and paste, not necessary.

File: pkg/ingester/ingester.go (hunk 4 above)
Comment (tomwilkie) [inline pkg/ingester/ingester.go]: +1

File: pkg/ingester/ingester_lifecycle_test.go (hunk 2 above)
Comment (tomwilkie) [inline pkg/ingester/ingester_lifecycle_test.go]: 2 x 100ms (the pause above).  For the success case, we need to wait at least 100ms, and I added another 100ms for scheduling and pauses.

File: pkg/ingester/ingester_lifecycle_test.go (hunk 1 above)
Comment (tomwilkie) [inline pkg/ingester/ingester_lifecycle_test.go]: We are trying to test that two things happen sequentially and not in parallel, so I've slowed one of them down, waited for the other to happen, and then checked the slowed down one hasn't happened.

I can't really think of a better way to test for this that gives a negative signal when they happen in the wrong order.

**Pull request merged**